packages/dx-react-core/src/plugin-based/action.jsx

```diff
@@ -50,12 +52,17 @@ export class Action extends React.PureComponent {
   }
 }
 
-Action.propTypes = {
+ActionBase.propTypes = {
   name: PropTypes.string.isRequired,
   action: PropTypes.func.isRequired,
-};
-
-Action.contextTypes = {
   [PLUGIN_HOST_CONTEXT]: PropTypes.object.isRequired,
   [POSITION_CONTEXT]: PropTypes.func.isRequired,
 };
+
+export const Action = withPropsFromContext({
```

Comment:
This function is called 4 times with same arguments and 1 time with a different second argument.
Its name implies that component is connected to a single context.

How about changing it so that it actually connects component to a single context?
```javascript
const TargetWithContext = withContext(SomeContext, name)(Target);
```

And then introducing `withHostAndPosition` function?
```javascript
const Action = withHostAndPosition(ActionBase);  // Same for Getter, Plugin, Template
```
```javascript
const TemplatePlaceholder = withHost(
  withContext(TemplateHostContext, TEMPLATE_HOST_CONTEXT)(
    TemplatePlaceholderBase
  )
);
```

Reply:
Done